cps/kni.c

```diff
@@ -44,6 +44,61 @@
 extern long init_module(void *, unsigned long, const char *);
 extern long delete_module(const char *, unsigned int);
 
+/*
+ * The minimum length of each update is a Netlink message header,
+ * a routing table message header, an IPv4 address for the route,
+ * an IPv4 address for the gateway, and a 32-bit interface ID.
+ *
+ */
+#define MIN_RTNL_UPDATE_LEN (sizeof(struct nlmsghdr) + sizeof(struct rtmsg) + \
+	2 * sizeof(struct in_addr) + sizeof(uint32_t))
+
+/* Maximum number of route updates that can fit in a Netlink message. */
+#define MAX_RTNL_UPDATES (MNL_SOCKET_BUFFER_SIZE / MIN_RTNL_UPDATE_LEN)
+
+struct route_update {
+	/* Type of route update: RTM_NEWROUTE or RTM_DELROUTE. */
+	int      type;
+
+	/* Address family of route: AF_INET or AF_INET6. */
+	int      family;
+
+	/*
+	 * Whether this update has all the fields and attributes
+	 * necessary to update the LPM table.
+	 */
+	int      valid;
+
+	/* Route prefix length. */
+	uint8_t  prefix_len;
+
+	/* Output interface index of route. */
+	uint32_t oif_index;
+
+	/* Route origin. See field rtm_protocol of struct rtmsg. */
+	uint8_t  rt_proto;
+
+	/* IP address of route. */
+	union {
+		struct in_addr  v4;
+		struct in6_addr v6;
+	} ip;
+
+	/* IP address of route gateway. */
+	union {
+		struct in_addr  v4;
+		struct in6_addr v6;
+	} gw;
+};
+
+struct route_updates {
+	/* Pointer to route updates. */
+	struct route_update *u;
```

Comment:
Because the definition of `MAX_RTNL_UPDATES` contains arithmetic and isn't a constant number, the compiler does not allow it. In other words, if you did this:

-   struct route_update *u;
+  struct route_update u[MAX_RTNL_UPDATES];

Then you get this error:

/home/cody/gatekeeper/cps/kni.c:96:22: error: variably modified ‘u’ at file scope
  struct route_update u[MAX_RTNL_UPDATES];

Reply:
What makes `MAX_RTNL_UPDATES` not a contant is `MNL_SOCKET_BUFFER_SIZE`, whose definition is `#define MNL_SOCKET_BUFFER_SIZE (getpagesize() < 8192L ? getpagesize() : 8192L)`. Thus, if we replace `MNL_SOCKET_BUFFER_SIZE` with `8192L` (i.e. 8KB), `MAX_RTNL_UPDATES` becomes a constant.

Reply:
That... makes a lot more sense. Thanks!

Reply:
It'd be nice to find a constant for `8192L`. Perhaps, the maximum page size?

Reply:
I wasn't able to find anything for that. I searched through the kernel headers too for 8192 with no luck.

Reply:
Let's hardcore it in this case.

Reply:
We end up not needing any of this.

By accepting the fact that Netlink packets may have multiple messages per packet, we limit the update routing by the number of _packets_ instead of the number of _updates_. This means that we can process however many packets we are allowed and for each message in the packets, add or delete the route as needed. We don't need to allocate room for X number of updates and pass that information back and forth.